crates/bevy_ecs/src/world/mod.rs

```diff
@@ -2635,6 +2635,212 @@ impl World {
                 .get_mut_by_id(component_id)
         }
     }
+
+    /// Retrieves a reference to the given `entity`'s [`Component`] of the given `type_id` using
+    /// reflection.
+    ///
+    /// Requires [`#[derive(Reflect)`](derive@bevy_reflect::Reflect) on the [`Component`] and
+    /// `app.register_type::<TheComponent>()` to have been called[^note-reflect-impl].
+    ///
+    /// If you want to call this with a [`ComponentId`], see [`World::components()`] and [`Components::get_id()`] to get
+    /// the corresponding [`TypeId`].
+    ///
+    ///
+    /// Also see the crate documentation for [`bevy_reflect`] for more information on
+    /// [`Reflect`] and bevy's reflection capabilities.
+    ///
+    /// # Errors
+    ///
+    /// See [`GetComponentReflectError`](error::GetComponentReflectError) for the possible errors and their descriptions.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use bevy_ecs::prelude::*;
+    /// use bevy_reflect::Reflect;
+    /// use std::any::TypeId;
+    ///
+    /// // define a `Component` and derive `Reflect` for it
+    /// #[derive(Component, Reflect)]
+    /// struct MyComponent;
+    ///
+    /// // create a `World` for this example
+    /// let mut world = World::new();
+    ///
+    /// // Note: This is usually handled by `App::register_type()`, but this example can not use `App`.
+    /// world.init_resource::<AppTypeRegistry>();
+    /// world.get_resource_mut::<AppTypeRegistry>().unwrap().write().register::<MyComponent>();
+    ///
+    /// // spawn an entity with a `MyComponent`
+    /// let entity = world.spawn(MyComponent).id();
+    ///
+    /// // retrieve a reflected reference to the entity's `MyComponent`
+    /// let comp_reflected: &dyn Reflect = world.get_reflect(entity, TypeId::of::<MyComponent>()).unwrap();
+    ///
+    /// // make sure we got the expected type
+    /// assert!(comp_reflected.is::<MyComponent>());
+    /// ```
+    ///
+    /// # Note
+    /// Requires the `bevy_reflect` feature (included in the default features).
+    ///
+    /// [`ReflectFromPtr`]: bevy_reflect::ReflectFromPtr
+    /// [`TypeData`]: bevy_reflect::TypeData
+    /// [`Reflect`]: bevy_reflect::Reflect
+    /// [`App::register_type`]: ../../bevy_app/struct.App.html#method.register_type
+    /// [^note-reflect-impl]: More specifically: Requires [`TypeData`] for [`ReflectFromPtr`] to be registered for the given `type_id`,
+    ///     which is automatically handled when deriving [`Reflect`] and calling [`App::register_type`].
+    #[inline]
+    #[cfg(feature = "bevy_reflect")]
+    pub fn get_reflect(
+        &self,
+        entity: Entity,
+        type_id: TypeId,
+    ) -> Result<&dyn bevy_reflect::Reflect, error::GetComponentReflectError> {
```

Comment:
Should this return a `Ref<'_, dyn bevy_reflect::Reflect>` so users can check for changes and such?

Reply:
Ah, good point. I went with modeling it similar to `get_by_id()`, but will see if it's easy to add the `Ref<>`.

Reply:
Ok, so, this is not suuper easy, and I'm not sure whether to do it in this PR.

The point is that `World::get_mut_by_id` returns a `MutUntyped`, but `World::get_by_id` only returns a `Ptr`. There is also no type `RefUntyped` (yet). One way to go about this could be to a) add `RefUntyped`, b) change `World::get_by_id` to return a `RefUntyped` instead of `Ptr` (possibly breaking change?), c) adapt `World:get_reflect` to return `Ref` from the internal `RefUntyped`, similar to how `World::get_reflect_mut` does it already.

I'd think this refactoring shouldn't go into this PR, but it would avoid having to also breakingly change `World::get_reflect` from `&dyn Reflect` to `Ref<'_, dyn Reflect>` later on. But imo that's not suuper important, as the change could be breaking for `World::get_by_id` anyway, so there would already be a breaking change.

I also looked into whether it's possible to (easily) directly return a `Ref` from `World::get_reflect` now, but it looked to be a bit beyond me. The only way to get the `Ticks` required to build the `Ref` seems to involve unsafe functions, like `get_components_and_ticks` (in `unsafe_world_cell.rs`).

But I'd be willing to take another look at this if someone can give me a rough outline/some guiding pointers on how best to do this, since it doesn't look too difficult either :)

Reply:
No worries! We can save this for a followup :)

Reply:
Ok, got it! :)